datastore/src/main/java/org/jboss/pnc/datastore/repositories/ProductMilestoneRepositoryImpl.java

```diff
@@ -30,4 +54,194 @@ public class ProductMilestoneRepositoryImpl extends AbstractRepository<ProductMi
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<Artifact> artifacts = query.from(Artifact.class);
+        Join<Artifact, BuildRecord> builtArtifacts = artifacts.join(Artifact_.buildRecord);
+        Join<BuildRecord, org.jboss.pnc.model.ProductMilestone> builtArtifactsMilestones = builtArtifacts
+                .join(BuildRecord_.productMilestone);
+
+        query.where(cb.equal(builtArtifactsMilestones.get(ProductMilestone_.id), id));
+        query.select(cb.count(artifacts.get(Artifact_.id)));
+
+        return entityManager.createQuery(query).getSingleResult();
+    }
+
+    @Override
+    public long countDeliveredArtifactsBuiltInThisMilestone(Integer id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<ProductMilestone> milestones = query.from(ProductMilestone.class);
+        SetJoin<ProductMilestone, Artifact> deliveredArtifacts = milestones.join(ProductMilestone_.deliveredArtifacts);
+        Path<ProductMilestone> builtDeliveredArtifactsMilestone = deliveredArtifacts.join(Artifact_.buildRecord)
+                .get(BuildRecord_.productMilestone);
+
+        // 1) we care about artifacts from the milestone given by id
+        // 2) only built delivered artifacts built in this milestone => implies built by this product
+        query.where(
+                cb.equal(milestones.get(ProductMilestone_.id), id),
+                cb.equal(builtDeliveredArtifactsMilestone.get(ProductMilestone_.id), id));
+        query.select(cb.count(deliveredArtifacts.get(Artifact_.id)));
+
+        return entityManager.createQuery(query).getSingleResult();
+    }
+
+    @Override
+    public long countDeliveredArtifactsBuiltInOtherMilestones(Integer id) {
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<Long> query = cb.createQuery(Long.class);
+
+        Root<ProductMilestone> milestones = query.from(ProductMilestone.class);
+        SetJoin<ProductMilestone, Artifact> deliveredArtifacts = milestones.join(ProductMilestone_.deliveredArtifacts);
+        Path<ProductMilestone> builtDeliveredArtifactsMilestone = deliveredArtifacts.get(Artifact_.buildRecord)
+                .get(BuildRecord_.productMilestone);
+
+        // 1) we care about artifacts from the milestone given by id
+        // 2) only built delivered artifacts from this product
+        // 3) only built delivered artifacts *not* built in this milestone
+        query.where(
+                cb.equal(milestones.get(ProductMilestone_.id), id),
+                cb.equal(
+                        builtDeliveredArtifactsMilestone.get(ProductMilestone_.productVersion)
+                                .get(ProductVersion_.product)
+                                .get(Product_.id),
+                        getProductIdByItsMilestone(id)),
```

Comment:
comparing the products directly in one query doesn't work?

Reply:
Done, see 68f63d5